### PR TITLE
Polarity

### LINKED ISSUE #213: Implement a new polarity classifier
# Issue description

Basically the same as https://github.com/METResearchGroup/social_agent_simulation_platform/issues/212 and https://github.com/METResearchGroup/social_agent_simulation_platform/issues/205, but for implementing a polarity ML classifier.

Other related resources to check out:
- https://github.com/METResearchGroup/social_agent_simulation_platform/blob/main/strategy_planning/2026-03-13_feature_extraction_api/FEATURE_EXTRACTION_API_PRD.md
- https://github.com/METResearchGroup/social_agent_simulation_platform/blob/main/strategy_planning/2026-03-13_feature_extraction_api/INTERN_PROJECT_CONTRACT.md

## Description

We want to track the polarity of posts. This is split into three labels, "positive", "negative", and" neutral".

## Model

Let's use `cardiffnlp/twitter-roberta-base-sentiment-latest `, a RoBERTa-base sentiment classifier trained on ~124M English tweets (2018–2021) and finetuned on TweetEval sentiment (3-way: negative/neutral/positive).

## Current requirements

Add a small module under `ml_tooling/` that provides a stable Python API.

The code should look like this:

```python
ml_tooling/polarity/
   models.py # put pydantic models here
   classifier.py # put classifier here
   examples_test.py # put examples here that can be run as a one-off script.
   tests/
      test_classifier.py # put pytest unit tests here. These will be run in CI.
```

### Public API

- `extract_polarity(text: str) -> PolarityLabel`
- `extract_polarity_batch(texts: list[str]) -> list[PolarityLabel]`
- Can implement batch like you did for https://github.com/METResearchGroup/social_agent_simulation_platform/issues/205.

### Output type

Use a Pydantic model with:

```python
class PolarityValue(enum): # or something like this, I forget the exact syntax, but create an enum to define the PolarityValue as positive/negative/neutral
   

class PolarityLabel:
   text_id: str
   text: str
   polarity_label: PolarityValue
   polarity_prob: float # probability of the highest label class
   label_timestamp: str # use lib/timestamp_utils.py to get the timestamp
```
### Basic implementation example

This is just a code snippet for how to use the model, to show how it works:

```python
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch
from scipy.special import softmax

MODEL_NAME = "cardiffnlp/twitter-roberta-base-sentiment-latest"

tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
model = AutoModelForSequenceClassification.from_pretrained(MODEL_NAME)

# Optional: simple tweet-style preprocessing
def preprocess(text: str) -> str:
    tokens = []
    for t in text.split(" "):
        if t.startswith("@") and len(t) > 1:
            t = "@user"
        if t.startswith("http"):
            t = "http"
        tokens.append(t)
    return " ".join(tokens)

text = "I love using this model on Twitter data!"
text_prep = preprocess(text)

encoded = tokenizer(text_prep, return_tensors="pt")
with torch.no_grad():
    output = model(**encoded)

scores = output.logits[0].numpy()
probs = softmax(scores)

labels = ["negative", "neutral", "positive"]
for label, p in zip(labels, probs):
    print(label, float(p))
```

### PR DESCRIPTION
**Pull request by dudu-theman**
**Problem**
We need a polarity classifier to classify the sentiment/polarity of a message

Fixes #213.

**Solution**
Create the polarity classifier that takes in a string/set of strings and outputs the most likely sentiment of the message + other metadata.

**Manual Verification**
Sanity check:
```
python ml_tooling/polarity/examples_test.py
```
Pytest:
```
pytest ml_tooling/polarity/tests/test_classifier.py
```

**State (after)**:
image

**Model Performance**:
- About 5x slower than emotion classifier
- About 3x slower than named entity recognition classifier.

## Summary by CodeRabbit

* **New Features**
  * Added sentiment (polarity) analysis supporting single and batch inputs; returns structured labels with predicted class, top probability, timestamp, and text identifier.
  * Exposed a default inference pipeline with input normalization and batch handling.

* **Tests**
  * Added unit tests for label conversion, single/batch behavior, and batch-size error cases; included an example/benchmark script demonstrating usage and throughput.

* **Chores**
  * Centralized ML dependencies into a shared extra and added an optional polarity extra.

### CODE REVIEW

File: ml_tooling/polarity/examples_test.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+
+if __name__ == "__main__":
+    polarity_model = track_init_time()
+    print("\n")
+
+    verify_diff_cases(polarity_model)
+    print("\n")
+
+    run_model_track_time(polarity_model)
```
[the remark this answers is not in the document]
Comment (dudu-theman) [reply ml_tooling/polarity/examples_test.py]: sounds good!

File: ml_tooling/polarity/classifier.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                t = "http"
+            tokens.append(t)
+        return " ".join(tokens)
+
+    def __call__(self, text: str | list[str]) -> PolarityResponse:
+        is_single = isinstance(text, str)
+
+        input_list = [text] if is_single else text
+
+        normalized_inputs = [self._preprocess(t) for t in input_list]
```
Comment (mark-torres10) [inline ml_tooling/polarity/classifier.py]: good job on the list comprehension! This is looking very much like professional Python code.

File: ml_tooling/polarity/classifier.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        encoded = self.tokenizer(
+            normalized_inputs, return_tensors="pt", padding=True, truncation=True
+        )
+
+        with torch.no_grad():
```
[the remark this answers is not in the document]
Comment (mark-torres10) [reply ml_tooling/polarity/classifier.py]: yes, that's on the right track. Have you experimented with what happens when you do/don't include the `with torch.no_grad()` to see what happens?

**Comment by dudu-theman** [comment]
TODO: refactor examples_test.py code into helpers across all ml_tooling. @mark-torres10  let me know if there's any other issues in this PR that I should address.
